maven-embedder/src/main/java/org/apache/maven/cli/transfer/FileSizeFormat.java

```diff
@@ -0,0 +1,160 @@
+ *
+ * @see <a href="https://en.wikipedia.org/wiki/Metric_prefix">https://en.wikipedia.org/wiki/Metric_prefix</a>
+ * @see <a href="https://en.wikipedia.org/wiki/Binary_prefix">https://en.wikipedia.org/wiki/Binary_prefix</a>
+ * @see <a
+ *      href="https://en.wikipedia.org/wiki/Octet_%28computing%29">https://en.wikipedia.org/wiki/Octet_(computing)</a>
+ */
+public class FileSizeFormat {
+    enum ScaleUnit {
+        BYTE {
+            @Override
+            public long bytes() {
+                return 1L;
+            }
+
+            @Override
+            public String symbol() {
+                return "B";
+            }
+        },
+        KILOBYTE {
+            @Override
+            public long bytes() {
+                return 1000L;
+            }
+
+            @Override
+            public String symbol() {
+                return "kB";
+            }
+        },
+        MEGABYTE {
+            @Override
+            public long bytes() {
+                return KILOBYTE.bytes() * KILOBYTE.bytes();
+            }
+
+            @Override
+            public String symbol() {
+                return "MB";
+            }
+        },
+        GIGABYTE {
+            @Override
+            public long bytes() {
+                return MEGABYTE.bytes() * KILOBYTE.bytes();
+            }
+            ;
+
+            @Override
+            public String symbol() {
+                return "GB";
+            }
+        };
+
+        public abstract long bytes();
+
+        public abstract String symbol();
+
+        public static ScaleUnit getScaleUnit(long size) {
+            if (size < 0L) {
+                throw new IllegalArgumentException("file size cannot be negative: " + size);
+            }
+
+            if (size >= GIGABYTE.bytes()) {
+                return GIGABYTE;
+            } else if (size >= MEGABYTE.bytes()) {
+                return MEGABYTE;
+            } else if (size >= KILOBYTE.bytes()) {
+                return KILOBYTE;
+            } else {
+                return BYTE;
+            }
+        }
+    }
+
+    private DecimalFormat smallFormat;
+    private DecimalFormat largeFormat;
+
+    public FileSizeFormat(Locale locale) {
+        smallFormat = new DecimalFormat("#0.0", new DecimalFormatSymbols(locale));
+        largeFormat = new DecimalFormat("###0", new DecimalFormatSymbols(locale));
+    }
+
+    public String format(long size) {
+        return format(size, ScaleUnit.getScaleUnit(size));
+    }
+
+    public String getScaleSymbol(long size) {
+        return ScaleUnit.getScaleUnit(size).symbol();
+    }
+
+    @SuppressWarnings("checkstyle:magicnumber")
+    private String format(long size, ScaleUnit unit) {
+        if (size < 0L) {
+            throw new IllegalArgumentException("file size cannot be negative: " + size);
+        }
+
+        double scaledSize = (double) size / unit.bytes();
+
+        if (unit == ScaleUnit.BYTE) {
+            return largeFormat.format(size);
```

Comment:
The goal is to avoid those calls.  
`NumberFormat` has methods that take a `StringBuilder`...
In `NumberFormat`, I have the following declaration:
```
    public final String format(long number) {
        return format(number, new StringBuffer(),
                      DontCareFieldPosition.INSTANCE).toString();
    }
```
so we should go straight to the method that takes a `StringBuffer` to avoid unneeded allocations + toString....

Reply:
Would have been nice, but the NumberFormat.format() method takes only StringBuffer as parameter, the synchronized StringBuilder. They are not interchangeable  and I am not sure we should switch StringBuilder for StringBuffer for that....

Reply:
What about `new Formatter(locale, builder).format("%3.f", size)`  ? i.e. do not use the `NumberFormat`, but `Formatter` ?

Reply:
Using a Formatter, the pattern string is passed as parameter to the format() method. Doing this the pattern is recompiled each method call and not reused. This generates a lot of garbage. A more optimised solution would be to avoid using generic Format or Formatter classes and implement a custom logic specific to this particular case. Some math logic to round the numbers as needed. This would reduce even more the memory allocation.

Reply:
@gnodet you can find the refactored `FileSizeFormat` class using Math logic in the last commit